// @lynx-js/web-elements/dist/XViewpagerNg/XViewpagerNg.js
import { __esDecorate, __runInitializers } from "tslib";
/*
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
*/
import { boostedQueueMicrotask, Component, } from '@lynx-js/web-elements-reactive';
import { CommonEventsAndMethods } from '../common/CommonEventsAndMethods.js';
import { XViewpagerNgEvents } from './XViewpagerNgEvents.js';
import { scrollContainerDom } from '../common/constants.js';
import { templateXViewpageNg } from '@lynx-js/web-elements-template';
let XViewpagerNg = (() => {
    let _classDecorators = [Component('x-viewpager-ng', [CommonEventsAndMethods, XViewpagerNgEvents], templateXViewpageNg)];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = HTMLElement;
    var XViewpagerNg = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            XViewpagerNg = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static notToFilterFalseAttributes = new Set([
            'allow-horizontal-gesture',
            'enable-scroll',
        ]);
        selectTab(params) {
            let { index, smooth } = params;
            if (typeof smooth === 'undefined')
                smooth = true;
            const scrollContainer = this.shadowRoot.children.item(2);
            const scrollLeft = scrollContainer.clientWidth * index;
            scrollContainer.scrollTo({
                left: scrollLeft,
                behavior: smooth ? 'smooth' : 'instant',
            });
        }
        connectedCallback() {
            const initialSelectIndex = this.getAttribute('select-index')
                || this.getAttribute('initial-select-index');
            if (initialSelectIndex !== null) {
                const selectIndex = Number(initialSelectIndex);
                const scrollContainer = this.shadowRoot.children.item(2);
                const scrollToInitialIndex = () => {
                    if (scrollContainer.clientWidth === 0) {
                        // In Safari, there is the potential race condition between the browser's layout and clientWidth calculate.
                        // So, we have to use requestAnimationFrame to ensure that the code runs after the browser's layout.
                        requestAnimationFrame(scrollToInitialIndex);
                    }
                    else {
                        this.selectTab({ index: selectIndex, smooth: false });
                    }
                };
                // The reason for using microtasks is that the width and height of the child element may not be rendered at this time, so it will not be able to scroll.
                boostedQueueMicrotask(scrollToInitialIndex);
            }
        }
        get [scrollContainerDom]() {
            return this;
        }
        static {
            __runInitializers(_classThis, _classExtraInitializers);
        }
    };
    return XViewpagerNg = _classThis;
})();
export { XViewpagerNg };
